### ISSUE TITLE
[CTA][Security] Qualys detect xss 

### ISSUE THREAD

**Issue by cedrox** [inboundnow/inbound-pro#171]
Hello, 
There is a security issue detected by Qualys Scan : Cross-site Scripting XSS.

Details :
Parameter | It has been detected by exploiting the parameter wp_lead_uid
The payloads section will display a list of tests that show how the param could have been exploited to collect the information

Authentication | In order to detect this vulnerability, no authentication has been required.
All the public url are impacted

Here the typical Payload 

Payload "'><qss%20a=@REQUESTID@>
Request GET [All URL]
#1 Cookie: **wp_lead_uid=%22'%3E%3Cqss%20a%3DX3036611716Y0Z%3E;** wordpress_test_cookie=WP+Cookie+check; search=1; s=1; lead_session=1;
inbound_referral_site=Direct Traffic; cat=3; aam_uuid=18484822287460117101663997602216214954; PHPSESSID=e8841f428c81d6fad8ba431c16f29863;
MSFPC=GUID=1e5afa6d046145deb2099c8d1f8c6bf4&HASH=1e5a&LV=201709&V=4&LU=1505991110179;
ARRAffinity=ffb17ed79e90f591213c78aaa374534f7a36f40308e00f8414d851603f954158; AMCV_EA76ADE95776D2EC7F000101%40AdobeOrg=-179204249%7CMCMID
%7C18023176558317847531707862135258802341%7CMCAAMLH-1506595910%7C9%7CMCAAMB-1506596446%7CcIBAx_aQzFEHcPoEv0GwcQ
%7CMCOPTOUT-1505998310s%7CNONE%7CMCAID%7CNONE%7CMCCIDH%7C433268694; AMCVS_EA76ADE95776D2EC7F000101%40AdobeOrg=1; 45531=;
45530=;
#2 Referer: [Home URL]

![image](https://user-images.githubusercontent.com/10764310/31454812-0644e3e6-aeb7-11e7-87ca-11a869500a9b.png)

Thanks

**Comment by atwellpub** [comment]
Very interesting thank you. I added in sanitations that should stop the possible exploitation. Will run updates soon.

**Comment by cedrox** [comment]
Hello @atwellpub , 
the sanitation mechnism seem not suffissant.
The return scan say that :

Group | Cross-Site Scripting
WASC | Improper Input Handling
150084 | Unencoded   characters


Details
--
Threat | The web application reflects potentially dangerous   characters such as single quotes, double quotes, and angle brackets. These   characters are commonly used for HTML injection attacks such as cross-site   scripting (XSS).
Impact | No exploit was determined for these reflected   characters. The input parameter should be manually analyzed to verify that no   other characters can be injected that would lead to an HTML injection (XSS)   vulnerability.
Solution | Review the reflected characters to ensure that they   are properly handled as defined by the web application's coding practice.   Typical solutions are to apply HTML encoding or percent encoding to the   characters depending on where they are placed in the HTML. For example, a   double quote might be encoded as " when displayed in a text node, but as   %22 when placed in the value of an href attribute.

Parameter | wp_lead_uid
Authentication | Not Required( Need Clarification)
Access Path | ****

#1 Request
--
Payload | "'><qss%20a=@REQUESTID@>
Request | GET
#1 Cookie:   wp_lead_uid=%22'%3E%3Cqss%20a%3DX3049546036Y0Z%3E; search=1; s=1;   lead_session=1; inbound_referral_site=Direct Traffic; cat=3;   ai_user=6et++\|2017-10-17T08:29:31.798Z;   PHPSESSID=3bc616fe39e08795f9a1ee8d76474d2f; MSFPC=GUID=ca2e89e4ce2e4be1aa97959c9420eb6d&HASH=ca2e&LV=201710&V=4&LU=1508228406118;   ARRAffinity=e54c3a3912a3f14a1d44ba162eca3d4aefc0a37c73b2ef83c2025fcf334c608b;
#2 Referer:   http://[websiteurl]/
#1 Response
comment: A   significant portion of the XSS test payload appeared in the web page, but the   page's DOM was not modified as expected for a successful exploit. This result   should be manually verified to determine its accuracy.          Response content-type: text/html          .js'></script>     <script type='text/javascript'>     /* <![CDATA[ */     var inbound_settings =   {"post_id":"cat_25","post_type":"post","variation_id":"0","ip_address":"172.20.0.1","wp_lead_data":{"lead_id":null,"lead_email":null,"lead_uid":"\\\"\\'><qss   a=X3049546036Y0Z>","lead_nonce":null},"admin_url":"https:\/\/[websiteurl]\/wp-admin\/admin-ajax.php","track_time":"2017\/10\/17   18:28:43","page_tracking":"off","search_tracking":"off","comment_tracking":"off","custom_mapping":[],"is_admin":"","

comment: A significant portion of the XSS test payload appeared   in the web page, but the page's DOM was not modified as expected for a   successful exploit. This result should be manually verified to determine its   accuracy.          Response content-type: text/html          iation.js'></script>     <script type='text/javascript'>     /* <![CDATA[ */     var inbound_settings =   {"post_id":"cat_25","post_type":"post","variation_id":"0","ip_address":"172.20.0.1","wp_lead_data":{"lead_id":null,"lead_email":null,"lead_uid":"<IMG   SRC=javascript:qss=7>","lead_nonce":null},"admin_url":"https:\/\/[websiteurl]\/wp-admin\/admin-ajax.php","track_time":"2017\/10\/17   17:12:40","page_tracking":"off","search_tracking":"off","comment_tracking":"off","custom_mapping":[],"is_admin":""
--

**Comment by atwellpub** [comment]
Hey @cedrox 

Do you understand what is happening? As it is now the report is a little over my head but the sanitations I added should prevent any PHP manipulation.

**Comment by cedrox** [comment]
Thanks for the follow up :)
I understand that the code bellow can be executed on client side... If you pass a value in the payload like
wp_lead_uid=%22'%3E%3Cqss%20a%3DX3049546036Y0Z%3E;
script can be excecuted as it's writen in 
                <![CDATA[ */
var inbound_settings = {"post_id":"cat_26","post_type":"post","variation_id":"0","ip_address":"1.1.1.1","wp_lead_data":{"lead_id":[malicious code],"lead_email":null,"lead_uid":null,"lead_nonce":null},"admin_url":"https:\/\/[siteurl]\/wp-admin\/admin-ajax.php","track_time":"2017\/10\/23 17:46:17","page_tracking":"off","search_tracking":"off","comment_tracking":"off","custom_mapping":[],"is_admin":"","ajax_nonce":"6738ebc33d"};
/* ]]> */

Maybe https://developer.wordpress.org/reference/functions/sanitize_text_field/ is not enough...

Thanks

**Comment by cedrox** [comment]
Here some input from security team : 
In addition to using sanitize text field, there should be encoding used when displaying data especially user input.

**Comment by atwellpub** [comment]
I don't think we ever will display the cookie data wp_lead_uid on the front end, so the value will never get rendered client side. But we do send that data to the PHP end every page load and that's were we perform the sanitation. It could be used to run PHP operations, but the sanitize_text_field() takes care of it at the backend. Is there a visual representation of the exploit in action I could use to better understand?

Kind regards and thank you for the continued attention I do appreciate it,
Hudson

**Comment by cedrox** [comment]
Answer from security team :
its better to have strong validation against user input, like checking for inputs if they meet expected format. Sanitize_text_field() is great start but better to ensure you have tighter validation especially if you can tie down the input to the expected range. **If you are convinced that site has all of it already let me know we can classify it as false positive.**

Thanks

**Comment by atwellpub** [comment]
I believe it's a false positive as this input is a background input. User has no access to it and if it's modified somehow we are sanitizing it on the backend when we process it. 

If I can be proven wrong it would be very educational and helpful for future development projects. But I think we're safe on this one.

**Comment by cedrox** [comment]
Thanks a lot @atwellpub for your follow up !!
